larpandoracontent/LArHelpers/LArMCParticleHelper.cc

```diff
@@ -216,6 +216,42 @@ const MCParticle *LArMCParticleHelper::GetMainMCParticle(const ParticleFlowObjec
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+const MCParticle *LArMCParticleHelper::GetBestMCParticle(const ParticleFlowObject *const pPfo)
+{
+    CaloHitList caloHitList;
+    LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_U, caloHitList);
+    LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_V, caloHitList);
+    LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_W, caloHitList);
+
+    const MCParticleWeightMap mcParticleWeightMap(GetMCParticleWeightMap(caloHitList));
+
+    float             bestWeight(0.f);
+    const MCParticle *pBestMCParticle(nullptr);
+
+    // Iterate through map in a reproducible order
+    MCParticleVector mcParticleVector;
+
+    for (const MCParticleWeightMap::value_type &mapEntry : mcParticleWeightMap)
+        mcParticleVector.push_back(mapEntry.first);
+
+    std::sort(mcParticleVector.begin(), mcParticleVector.end(), PointerLessThan<MCParticle>());
+
+    for (const MCParticle *const pMCParticle : mcParticleVector)
+    {
+        const float currentWeight(mcParticleWeightMap.at(pMCParticle));
+
+        if (currentWeight > bestWeight)
+        {
+            pBestMCParticle = pMCParticle;
+            bestWeight      = currentWeight;
+        }
+    }
+
+    return pBestMCParticle; // can be nullptr
```

Comment:
I think our convention is to raise a StatusCodeException, with STATUS_CODE_NOT_FOUND, rather than asking the algorithm to deal with a nullptr. Should try to make sure this approach is adopted throughout, rather than risking dereferencing nullptr one day!